From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Diego Chagas <drochachagas@gmail.com>
Date: Wed, 7 Oct 2026 20:31:18 -0300
Subject: [PATCH] Remove tool: paint over an object to remove it (AI)

Photoshop's Remove tool (GimpRemoveTool, a GimpDrawTool), in the J group:
paint over an object, the strokes shown in Photoshop's see-through pink
with the brush outline under the pointer; on release (or Enter, without
Tool Options' Remove after each stroke; Escape clears) it runs GIMPhoto's
plug-in procedure gimphoto-remove with the strokes as JSON and the model
(LaMa, or FLUX.2 klein with Generative fill), from an idle callback once
the event is done, one removal at a time, the image, display and drawable
kept alive until it returns. [ and ] change the brush size
(tools-remove-size-set).

diff --git a/app/actions/tools-actions.c b/app/actions/tools-actions.c
index d2b02e64e3..4bd5cbeb37 100644
--- a/app/actions/tools-actions.c
+++ b/app/actions/tools-actions.c
@@ -271,6 +271,15 @@ static const GimpEnumActionEntry tools_paint_select_size_actions[] =
     NULL }
 };
 
+static const GimpEnumActionEntry tools_remove_size_actions[] =
+{
+  { "tools-remove-size-set",
+    GIMP_ICON_TOOL_HEAL,
+    "Set Remove Tool Brush Size", NULL, { NULL }, NULL,
+    GIMP_ACTION_SELECT_SET, TRUE,
+    NULL }
+};
+
 static const GimpEnumActionEntry tools_transform_preview_opacity_actions[] =
 {
   { "tools-transform-preview-opacity-set", GIMP_ICON_TOOL_PERSPECTIVE,
@@ -763,6 +772,11 @@ tools_actions_setup (GimpActionGroup *group)
                                       G_N_ELEMENTS (tools_paint_select_size_actions),
                                       tools_paint_select_size_cmd_callback);
 
+  gimp_action_group_add_enum_actions (group, NULL,
+                                      tools_remove_size_actions,
+                                      G_N_ELEMENTS (tools_remove_size_actions),
+                                      tools_remove_size_cmd_callback);
+
   gimp_action_group_add_enum_actions (group, NULL,
                                       tools_transform_preview_opacity_actions,
                                       G_N_ELEMENTS (tools_transform_preview_opacity_actions),
diff --git a/app/actions/tools-commands.c b/app/actions/tools-commands.c
index c8f1582a6f..3952b3eb4f 100644
--- a/app/actions/tools-commands.c
+++ b/app/actions/tools-commands.c
@@ -45,6 +45,7 @@
 #include "tools/gimpcoloroptions.h"
 #include "tools/gimpforegroundselectoptions.h"
 #include "tools/gimppaintselectoptions.h"
+#include "tools/gimpremoveoptions.h"
 #include "tools/gimprectangleoptions.h"
 #include "tools/gimptool.h"
 #include "tools/gimptoolcontrol.h"
@@ -613,6 +614,30 @@ tools_paint_select_size_cmd_callback (GimpAction *action,
     }
 }
 
+void
+tools_remove_size_cmd_callback (GimpAction *action,
+                                GVariant   *value,
+                                gpointer    data)
+{
+  GimpContext          *context;
+  GimpToolInfo         *tool_info;
+  GimpActionSelectType  select_type;
+  return_if_no_context (context, data);
+
+  select_type = (GimpActionSelectType) g_variant_get_int32 (value);
+
+  tool_info = gimp_context_get_tool (context);
+
+  if (tool_info && GIMP_IS_REMOVE_OPTIONS (tool_info->tool_options))
+    {
+      action_select_property (select_type,
+                              action_data_get_display (data),
+                              G_OBJECT (tool_info->tool_options),
+                              "size",
+                              1.0, 5.0, 50.0, 0.1, FALSE);
+    }
+}
+
 void
 tools_transform_preview_opacity_cmd_callback (GimpAction *action,
                                               GVariant   *value,
diff --git a/app/actions/tools-commands.h b/app/actions/tools-commands.h
index 01609b272a..5be4d5eaf5 100644
--- a/app/actions/tools-commands.h
+++ b/app/actions/tools-commands.h
@@ -88,6 +88,9 @@ void   tools_fg_select_brush_size_cmd_callback      (GimpAction *action,
 void   tools_paint_select_size_cmd_callback         (GimpAction *action,
                                                      GVariant   *value,
                                                      gpointer    data);
+void   tools_remove_size_cmd_callback               (GimpAction *action,
+                                                     GVariant   *value,
+                                                     gpointer    data);
 
 void   tools_transform_preview_opacity_cmd_callback (GimpAction *action,
                                                      GVariant   *value,
diff --git a/app/tools/gimp-tools.c b/app/tools/gimp-tools.c
index aa08c71fb4..b7a9bcf292 100644
--- a/app/tools/gimp-tools.c
+++ b/app/tools/gimp-tools.c
@@ -71,6 +71,7 @@
 #include "gimppathtool.h"
 #include "gimpshapetool.h"
 #include "gimpobjectselecttool.h"
+#include "gimpremovetool.h"
 #include "gimpmagnifytool.h"
 #include "gimpmeasuretool.h"
 #include "gimpmovetool.h"
@@ -178,6 +179,7 @@ gimp_tools_init (Gimp *gimp)
     gimp_text_tool_register,
     gimp_shape_tool_register,  /* GIMPhoto */
     gimp_object_select_tool_register,  /* GIMPhoto */
+    gimp_remove_tool_register,         /* GIMPhoto */
     gimp_bucket_fill_tool_register,
     gimp_gradient_tool_register,
     gimp_pencil_tool_register,
diff --git a/app/tools/gimpremoveoptions.c b/app/tools/gimpremoveoptions.c
new file mode 100644
index 0000000000..502b69ea10
--- /dev/null
+++ b/app/tools/gimpremoveoptions.c
@@ -0,0 +1,171 @@
+/* GIMP - The GNU Image Manipulation Program
+ * Copyright (C) 1995 Spencer Kimball and Peter Mattis
+ *
+ * gimpremoveoptions.c
+ * Copyright (C) 2026 GIMPhoto contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+
+#include <gegl.h>
+#include <gtk/gtk.h>
+
+#include "libgimpbase/gimpbase.h"
+#include "libgimpconfig/gimpconfig.h"
+#include "libgimpwidgets/gimpwidgets.h"
+
+#include "tools-types.h"
+
+#include "widgets/gimpwidgets-constructors.h"
+
+#include "gimpremoveoptions.h"
+#include "gimptooloptions-gui.h"
+
+#include "gimp-intl.h"
+
+
+enum
+{
+  PROP_0,
+  PROP_SIZE,
+  PROP_GENERATIVE,
+  PROP_REMOVE_AFTER_STROKE
+};
+
+
+static void   gimp_remove_options_set_property (GObject      *object,
+                                                guint         property_id,
+                                                const GValue *value,
+                                                GParamSpec   *pspec);
+static void   gimp_remove_options_get_property (GObject      *object,
+                                                guint         property_id,
+                                                GValue       *value,
+                                                GParamSpec   *pspec);
+
+
+G_DEFINE_TYPE (GimpRemoveOptions, gimp_remove_options,
+               GIMP_TYPE_TOOL_OPTIONS)
+
+
+static void
+gimp_remove_options_class_init (GimpRemoveOptionsClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->set_property = gimp_remove_options_set_property;
+  object_class->get_property = gimp_remove_options_get_property;
+
+  GIMP_CONFIG_PROP_INT (object_class, PROP_SIZE,
+                        "size",
+                        _("Size"),
+                        _("Size of the brush that marks what to remove"),
+                        1, 1000, 50,
+                        GIMP_PARAM_STATIC_STRINGS);
+
+  GIMP_CONFIG_PROP_BOOLEAN (object_class, PROP_GENERATIVE,
+                            "generative",
+                            _("Generative fill (slower)"),
+                            _("Fill with FLUX.2 klein (about 30 seconds), which "
+                              "redraws the area, instead of LaMa (about a "
+                              "second), which continues what is around it"),
+                            FALSE,
+                            GIMP_PARAM_STATIC_STRINGS);
+
+  GIMP_CONFIG_PROP_BOOLEAN (object_class, PROP_REMOVE_AFTER_STROKE,
+                            "remove-after-stroke",
+                            _("Remove after each stroke"),
+                            _("Remove when the button is released; otherwise "
+                              "strokes add up until Enter (Escape clears them)"),
+                            TRUE,
+                            GIMP_PARAM_STATIC_STRINGS);
+}
+
+static void
+gimp_remove_options_init (GimpRemoveOptions *options)
+{
+}
+
+static void
+gimp_remove_options_set_property (GObject      *object,
+                                  guint         property_id,
+                                  const GValue *value,
+                                  GParamSpec   *pspec)
+{
+  GimpRemoveOptions *options = GIMP_REMOVE_OPTIONS (object);
+
+  switch (property_id)
+    {
+    case PROP_SIZE:
+      options->size = g_value_get_int (value);
+      break;
+    case PROP_GENERATIVE:
+      options->generative = g_value_get_boolean (value);
+      break;
+    case PROP_REMOVE_AFTER_STROKE:
+      options->remove_after_stroke = g_value_get_boolean (value);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
+      break;
+    }
+}
+
+static void
+gimp_remove_options_get_property (GObject    *object,
+                                  guint       property_id,
+                                  GValue     *value,
+                                  GParamSpec *pspec)
+{
+  GimpRemoveOptions *options = GIMP_REMOVE_OPTIONS (object);
+
+  switch (property_id)
+    {
+    case PROP_SIZE:
+      g_value_set_int (value, options->size);
+      break;
+    case PROP_GENERATIVE:
+      g_value_set_boolean (value, options->generative);
+      break;
+    case PROP_REMOVE_AFTER_STROKE:
+      g_value_set_boolean (value, options->remove_after_stroke);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
+      break;
+    }
+}
+
+GtkWidget *
+gimp_remove_options_gui (GimpToolOptions *tool_options)
+{
+  GObject   *config = G_OBJECT (tool_options);
+  GtkWidget *vbox   = gimp_tool_options_gui (tool_options);
+  GtkWidget *scale;
+  GtkWidget *button;
+
+  scale = gimp_prop_spin_scale_new (config, "size", 1.0, 10.0, 0);
+  gimp_spin_scale_set_scale_limits (GIMP_SPIN_SCALE (scale), 1.0, 500.0);
+  gimp_spin_scale_set_gamma (GIMP_SPIN_SCALE (scale), 1.7);
+  gtk_box_pack_start (GTK_BOX (vbox), scale, FALSE, FALSE, 0);
+
+  button = gimp_prop_check_button_new (config, "remove-after-stroke", NULL);
+  gtk_box_pack_start (GTK_BOX (vbox), button, FALSE, FALSE, 0);
+
+  button = gimp_prop_check_button_new (config, "generative", NULL);
+  gtk_box_pack_start (GTK_BOX (vbox), button, FALSE, FALSE, 0);
+
+  return vbox;
+}
diff --git a/app/tools/gimpremoveoptions.h b/app/tools/gimpremoveoptions.h
new file mode 100644
index 0000000000..eea63695f3
--- /dev/null
+++ b/app/tools/gimpremoveoptions.h
@@ -0,0 +1,54 @@
+/* GIMP - The GNU Image Manipulation Program
+ * Copyright (C) 1995 Spencer Kimball and Peter Mattis
+ *
+ * gimpremoveoptions.h
+ * Copyright (C) 2026 GIMPhoto contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "core/gimptooloptions.h"
+
+
+#define GIMP_TYPE_REMOVE_OPTIONS            (gimp_remove_options_get_type ())
+#define GIMP_REMOVE_OPTIONS(obj)            (G_TYPE_CHECK_INSTANCE_CAST ((obj), GIMP_TYPE_REMOVE_OPTIONS, GimpRemoveOptions))
+#define GIMP_REMOVE_OPTIONS_CLASS(klass)    (G_TYPE_CHECK_CLASS_CAST ((klass), GIMP_TYPE_REMOVE_OPTIONS, GimpRemoveOptionsClass))
+#define GIMP_IS_REMOVE_OPTIONS(obj)         (G_TYPE_CHECK_INSTANCE_TYPE ((obj), GIMP_TYPE_REMOVE_OPTIONS))
+#define GIMP_IS_REMOVE_OPTIONS_CLASS(klass) (G_TYPE_CHECK_CLASS_TYPE ((klass), GIMP_TYPE_REMOVE_OPTIONS))
+#define GIMP_REMOVE_OPTIONS_GET_CLASS(obj)  (G_TYPE_INSTANCE_GET_CLASS ((obj), GIMP_TYPE_REMOVE_OPTIONS, GimpRemoveOptionsClass))
+
+
+typedef struct _GimpRemoveOptions      GimpRemoveOptions;
+typedef struct _GimpRemoveOptionsClass GimpRemoveOptionsClass;
+
+struct _GimpRemoveOptions
+{
+  GimpToolOptions  parent_instance;
+
+  gint             size;
+  gboolean         generative;
+  gboolean         remove_after_stroke;
+};
+
+struct _GimpRemoveOptionsClass
+{
+  GimpToolOptionsClass  parent_class;
+};
+
+
+GType       gimp_remove_options_get_type (void) G_GNUC_CONST;
+
+GtkWidget * gimp_remove_options_gui      (GimpToolOptions *tool_options);
diff --git a/app/tools/gimpremovetool.c b/app/tools/gimpremovetool.c
new file mode 100644
index 0000000000..f33665fa94
--- /dev/null
+++ b/app/tools/gimpremovetool.c
@@ -0,0 +1,575 @@
+/* GIMP - The GNU Image Manipulation Program
+ * Copyright (C) 1995 Spencer Kimball and Peter Mattis
+ *
+ * gimpremovetool.c
+ * Copyright (C) 2026 GIMPhoto contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+/*  GIMPhoto: Photoshop's Remove tool. Paint over an object (the strokes
+ *  shown pink); on release (or Enter, without "Remove after each stroke")
+ *  the plug-in procedure gimphoto-remove removes it with the local AI
+ *  (LaMa; FLUX.2 klein with Generative fill).
+ */
+
+#include "config.h"
+
+#include <gegl.h>
+#include <gtk/gtk.h>
+#include <gdk/gdkkeysyms.h>
+
+#include "libgimpbase/gimpbase.h"
+#include "libgimpmath/gimpmath.h"
+#include "libgimpwidgets/gimpwidgets.h"
+
+#include "tools-types.h"
+
+#include "core/gimp.h"
+#include "core/gimpdrawable.h"
+#include "core/gimpimage.h"
+#include "core/gimpparamspecs.h"
+#include "core/gimpprogress.h"
+
+#include "pdb/gimppdb.h"
+
+#include "widgets/gimphelp-ids.h"
+
+#include "display/gimpcanvaspen.h"
+#include "display/gimpdisplay.h"
+#include "display/gimpdisplayshell.h"
+
+#include "gimpremoveoptions.h"
+#include "gimpremovetool.h"
+#include "gimptoolcontrol.h"
+
+#include "gimp-intl.h"
+
+
+/*  the plug-in procedure that removes (plugins/generative-fill)  */
+#define REMOVE_PROC "gimphoto-remove"
+
+
+typedef struct
+{
+  gint    size;
+  GArray *points;  /*  GimpVector2, image coordinates  */
+} RemoveStroke;
+
+
+static void       gimp_remove_tool_finalize       (GObject               *object);
+
+static void       gimp_remove_tool_control        (GimpTool              *tool,
+                                                   GimpToolAction         action,
+                                                   GimpDisplay           *display);
+static void       gimp_remove_tool_button_press   (GimpTool              *tool,
+                                                   const GimpCoords      *coords,
+                                                   guint32                time,
+                                                   GdkModifierType        state,
+                                                   GimpButtonPressType    press_type,
+                                                   GimpDisplay           *display);
+static void       gimp_remove_tool_button_release (GimpTool              *tool,
+                                                   const GimpCoords      *coords,
+                                                   guint32                time,
+                                                   GdkModifierType        state,
+                                                   GimpButtonReleaseType  release_type,
+                                                   GimpDisplay           *display);
+static void       gimp_remove_tool_motion         (GimpTool              *tool,
+                                                   const GimpCoords      *coords,
+                                                   guint32                time,
+                                                   GdkModifierType        state,
+                                                   GimpDisplay           *display);
+static gboolean   gimp_remove_tool_key_press      (GimpTool              *tool,
+                                                   GdkEventKey           *kevent,
+                                                   GimpDisplay           *display);
+static void       gimp_remove_tool_oper_update    (GimpTool              *tool,
+                                                   const GimpCoords      *coords,
+                                                   GdkModifierType        state,
+                                                   gboolean               proximity,
+                                                   GimpDisplay           *display);
+
+static void       gimp_remove_tool_draw           (GimpDrawTool          *draw_tool);
+
+static void       gimp_remove_tool_clear          (GimpRemoveTool        *remove);
+static void       gimp_remove_tool_start          (GimpRemoveTool        *remove,
+                                                   GimpDisplay           *display);
+static gboolean   gimp_remove_tool_run_idle       (gpointer               data);
+
+
+G_DEFINE_TYPE (GimpRemoveTool, gimp_remove_tool, GIMP_TYPE_DRAW_TOOL)
+
+#define parent_class gimp_remove_tool_parent_class
+
+
+void
+gimp_remove_tool_register (GimpToolRegisterCallback  callback,
+                           gpointer                  data)
+{
+  (* callback) (GIMP_TYPE_REMOVE_TOOL,
+                GIMP_TYPE_REMOVE_OPTIONS,
+                gimp_remove_options_gui,
+                0,
+                "gimp-remove-tool",
+                _("Remove"),
+                _("Remove Tool: Paint over an object to remove it (AI)"),
+                N_("_Remove"), NULL,
+                NULL, GIMP_HELP_TOOL_HEAL,
+                "gimphoto-remove",
+                data);
+}
+
+static void
+gimp_remove_tool_class_init (GimpRemoveToolClass *klass)
+{
+  GObjectClass      *object_class    = G_OBJECT_CLASS (klass);
+  GimpToolClass     *tool_class      = GIMP_TOOL_CLASS (klass);
+  GimpDrawToolClass *draw_tool_class = GIMP_DRAW_TOOL_CLASS (klass);
+
+  object_class->finalize     = gimp_remove_tool_finalize;
+
+  tool_class->control        = gimp_remove_tool_control;
+  tool_class->button_press   = gimp_remove_tool_button_press;
+  tool_class->button_release = gimp_remove_tool_button_release;
+  tool_class->motion         = gimp_remove_tool_motion;
+  tool_class->key_press      = gimp_remove_tool_key_press;
+  tool_class->oper_update    = gimp_remove_tool_oper_update;
+
+  draw_tool_class->draw      = gimp_remove_tool_draw;
+}
+
+static void
+remove_stroke_free (gpointer data)
+{
+  RemoveStroke *stroke = data;
+
+  g_array_free (stroke->points, TRUE);
+  g_free (stroke);
+}
+
+static void
+gimp_remove_tool_init (GimpRemoveTool *remove)
+{
+  GimpTool *tool = GIMP_TOOL (remove);
+
+  gimp_tool_control_set_motion_mode   (tool->control, GIMP_MOTION_MODE_EXACT);
+  gimp_tool_control_set_scroll_lock   (tool->control, FALSE);
+  gimp_tool_control_set_precision     (tool->control,
+                                       GIMP_CURSOR_PRECISION_SUBPIXEL);
+  gimp_tool_control_set_tool_cursor   (tool->control,
+                                       GIMP_TOOL_CURSOR_PAINTBRUSH);
+  gimp_tool_control_set_action_size   (tool->control,
+                                       "tools-remove-size-set");
+
+  remove->strokes = g_ptr_array_new_with_free_func (remove_stroke_free);
+}
+
+static void
+gimp_remove_tool_finalize (GObject *object)
+{
+  GimpRemoveTool *remove = GIMP_REMOVE_TOOL (object);
+
+  g_clear_pointer (&remove->strokes, g_ptr_array_unref);
+  g_clear_pointer (&remove->stroke, remove_stroke_free);
+  g_clear_pointer (&remove->json, g_free);
+
+  G_OBJECT_CLASS (parent_class)->finalize (object);
+}
+
+static void
+gimp_remove_tool_control (GimpTool       *tool,
+                          GimpToolAction  action,
+                          GimpDisplay    *display)
+{
+  if (action == GIMP_TOOL_ACTION_HALT)
+    gimp_remove_tool_clear (GIMP_REMOVE_TOOL (tool));
+
+  GIMP_TOOL_CLASS (parent_class)->control (tool, action, display);
+}
+
+static void
+gimp_remove_tool_add_point (GimpRemoveTool   *remove,
+                            const GimpCoords *coords)
+{
+  GimpVector2 point = { coords->x, coords->y };
+
+  g_array_append_val (((RemoveStroke *) remove->stroke)->points, point);
+}
+
+static void
+gimp_remove_tool_button_press (GimpTool            *tool,
+                               const GimpCoords    *coords,
+                               guint32              time,
+                               GdkModifierType      state,
+                               GimpButtonPressType  press_type,
+                               GimpDisplay         *display)
+{
+  GimpRemoveTool    *remove  = GIMP_REMOVE_TOOL (tool);
+  GimpRemoveOptions *options = GIMP_REMOVE_TOOL_GET_OPTIONS (tool);
+  GimpImage         *image   = gimp_display_get_image (display);
+  GList             *drawables;
+  RemoveStroke      *stroke;
+
+  /*  the plug-in runs in a nested main loop: one removal at a time  */
+  if (remove->running || press_type != GIMP_BUTTON_PRESS_NORMAL)
+    return;
+
+  drawables = gimp_image_get_selected_drawables (image);
+
+  if (! drawables)
+    {
+      gimp_tool_message_literal (tool, display,
+                                 _("Select the layer to remove from."));
+      return;
+    }
+
+  g_list_free (drawables);
+
+  if (tool->display && tool->display != display)
+    gimp_tool_control (tool, GIMP_TOOL_ACTION_HALT, tool->display);
+
+  stroke         = g_new0 (RemoveStroke, 1);
+  stroke->size   = options->size;
+  stroke->points = g_array_new (FALSE, FALSE, sizeof (GimpVector2));
+  remove->stroke = stroke;
+  gimp_remove_tool_add_point (remove, coords);
+
+  tool->display = display;
+  gimp_tool_control_activate (tool->control);
+
+  gimp_draw_tool_pause (GIMP_DRAW_TOOL (tool));
+  if (! gimp_draw_tool_is_active (GIMP_DRAW_TOOL (tool)))
+    gimp_draw_tool_start (GIMP_DRAW_TOOL (tool), display);
+  gimp_draw_tool_resume (GIMP_DRAW_TOOL (tool));
+}
+
+static void
+gimp_remove_tool_motion (GimpTool         *tool,
+                         const GimpCoords *coords,
+                         guint32           time,
+                         GdkModifierType   state,
+                         GimpDisplay      *display)
+{
+  GimpRemoveTool *remove = GIMP_REMOVE_TOOL (tool);
+  GArray         *points;
+  GimpVector2    *last;
+
+  if (! remove->stroke)
+    return;
+
+  points = ((RemoveStroke *) remove->stroke)->points;
+  last   = &g_array_index (points, GimpVector2, points->len - 1);
+
+  gimp_draw_tool_pause (GIMP_DRAW_TOOL (tool));
+
+  remove->last_pos.x = coords->x;
+  remove->last_pos.y = coords->y;
+
+  if (fabs (coords->x - last->x) >= 1.0 || fabs (coords->y - last->y) >= 1.0)
+    gimp_remove_tool_add_point (remove, coords);
+
+  gimp_draw_tool_resume (GIMP_DRAW_TOOL (tool));
+}
+
+static void
+gimp_remove_tool_button_release (GimpTool              *tool,
+                                 const GimpCoords      *coords,
+                                 guint32                time,
+                                 GdkModifierType        state,
+                                 GimpButtonReleaseType  release_type,
+                                 GimpDisplay           *display)
+{
+  GimpRemoveTool    *remove  = GIMP_REMOVE_TOOL (tool);
+  GimpRemoveOptions *options = GIMP_REMOVE_TOOL_GET_OPTIONS (tool);
+
+  gimp_tool_control_halt (tool->control);
+
+  if (! remove->stroke)
+    return;
+
+  gimp_draw_tool_pause (GIMP_DRAW_TOOL (tool));
+
+  if (release_type == GIMP_BUTTON_RELEASE_CANCEL)
+    g_clear_pointer (&remove->stroke, remove_stroke_free);
+  else
+    g_ptr_array_add (remove->strokes, g_steal_pointer (&remove->stroke));
+
+  gimp_draw_tool_resume (GIMP_DRAW_TOOL (tool));
+
+  if (options->remove_after_stroke && remove->strokes->len > 0)
+    gimp_remove_tool_start (remove, display);
+}
+
+static gboolean
+gimp_remove_tool_key_press (GimpTool    *tool,
+                            GdkEventKey *kevent,
+                            GimpDisplay *display)
+{
+  GimpRemoveTool *remove = GIMP_REMOVE_TOOL (tool);
+
+  if (display != tool->display || remove->running)
+    return FALSE;
+
+  switch (kevent->keyval)
+    {
+    case GDK_KEY_Return:
+    case GDK_KEY_KP_Enter:
+    case GDK_KEY_ISO_Enter:
+      if (remove->strokes->len > 0)
+        {
+          gimp_remove_tool_start (remove, display);
+          return TRUE;
+        }
+      break;
+
+    case GDK_KEY_Escape:
+      if (remove->strokes->len > 0)
+        {
+          gimp_draw_tool_pause (GIMP_DRAW_TOOL (tool));
+          g_ptr_array_set_size (remove->strokes, 0);
+          gimp_draw_tool_resume (GIMP_DRAW_TOOL (tool));
+          return TRUE;
+        }
+      break;
+
+    default:
+      break;
+    }
+
+  return FALSE;
+}
+
+static void
+gimp_remove_tool_oper_update (GimpTool         *tool,
+                              const GimpCoords *coords,
+                              GdkModifierType   state,
+                              gboolean          proximity,
+                              GimpDisplay      *display)
+{
+  GimpRemoveTool *remove    = GIMP_REMOVE_TOOL (tool);
+  GimpDrawTool   *draw_tool = GIMP_DRAW_TOOL (tool);
+
+  if (proximity)
+    {
+      gimp_draw_tool_pause (draw_tool);
+
+      if (! tool->display || display == tool->display)
+        {
+          remove->last_pos.x = coords->x;
+          remove->last_pos.y = coords->y;
+        }
+
+      if (! gimp_draw_tool_is_active (draw_tool))
+        gimp_draw_tool_start (draw_tool, display);
+
+      gimp_draw_tool_resume (draw_tool);
+    }
+  else if (gimp_draw_tool_is_active (draw_tool) &&
+           remove->strokes->len == 0 && ! remove->stroke)
+    {
+      gimp_draw_tool_stop (draw_tool);
+    }
+}
+
+static void
+gimp_remove_tool_draw_stroke (GimpDrawTool *draw_tool,
+                              RemoveStroke *stroke,
+                              GeglColor    *color)
+{
+  GimpDisplayShell *shell = gimp_display_get_shell (draw_tool->display);
+  GArray           *points = stroke->points;
+  GimpVector2       dab[2];
+  GimpArray        *array;
+  GimpCanvasItem   *item;
+
+  if (points->len == 1)
+    {
+      /*  a click: a dab, as a stroke of two equal points  */
+      dab[0] = dab[1] = g_array_index (points, GimpVector2, 0);
+      array  = gimp_array_new ((const guint8 *) dab, sizeof (dab), TRUE);
+    }
+  else
+    {
+      array = gimp_array_new ((const guint8 *) points->data,
+                              points->len * sizeof (GimpVector2), TRUE);
+    }
+
+  item = g_object_new (GIMP_TYPE_CANVAS_PEN,
+                       "shell",  shell,
+                       "points", array,
+                       "color",  color,
+                       "width",  MAX (1, (gint) (stroke->size * shell->scale_y)),
+                       NULL);
+  gimp_draw_tool_add_item (draw_tool, item);
+
+  g_object_unref (item);
+  gimp_array_free (array);
+}
+
+static void
+gimp_remove_tool_draw (GimpDrawTool *draw_tool)
+{
+  GimpRemoveTool    *remove  = GIMP_REMOVE_TOOL (draw_tool);
+  GimpRemoveOptions *options = GIMP_REMOVE_TOOL_GET_OPTIONS (draw_tool);
+  GeglColor         *pink    = gegl_color_new (NULL);
+  guint              i;
+
+  /*  Photoshop's pink, see-through  */
+  gegl_color_set_rgba (pink, 1.0, 0.25, 0.6, 0.5);
+
+  for (i = 0; i < remove->strokes->len; i++)
+    gimp_remove_tool_draw_stroke (draw_tool,
+                                  g_ptr_array_index (remove->strokes, i),
+                                  pink);
+
+  if (remove->stroke)
+    gimp_remove_tool_draw_stroke (draw_tool, remove->stroke, pink);
+
+  g_object_unref (pink);
+
+  /*  the brush outline under the pointer  */
+  gimp_draw_tool_add_arc (draw_tool, FALSE,
+                          remove->last_pos.x - options->size / 2.0,
+                          remove->last_pos.y - options->size / 2.0,
+                          options->size, options->size,
+                          0.0, 2.0 * G_PI);
+}
+
+static void
+gimp_remove_tool_clear (GimpRemoveTool *remove)
+{
+  GimpDrawTool *draw_tool = GIMP_DRAW_TOOL (remove);
+
+  if (gimp_draw_tool_is_active (draw_tool))
+    gimp_draw_tool_stop (draw_tool);
+
+  g_clear_pointer (&remove->stroke, remove_stroke_free);
+  g_ptr_array_set_size (remove->strokes, 0);
+
+  GIMP_TOOL (remove)->display = NULL;
+}
+
+/*  JSON for the plug-in: [{"size": n, "points": [x0, y0, x1, y1, ...]}]  */
+static gchar *
+gimp_remove_tool_strokes_json (GimpRemoveTool *remove)
+{
+  GString *json = g_string_new ("[");
+  gchar    buf[G_ASCII_DTOSTR_BUF_SIZE];
+  guint    i, j;
+
+  for (i = 0; i < remove->strokes->len; i++)
+    {
+      RemoveStroke *stroke = g_ptr_array_index (remove->strokes, i);
+
+      g_string_append_printf (json, "%s{\"size\": %d, \"points\": [",
+                              i ? ", " : "", stroke->size);
+
+      for (j = 0; j < stroke->points->len; j++)
+        {
+          GimpVector2 *point = &g_array_index (stroke->points, GimpVector2, j);
+
+          g_string_append (json, j ? ", " : "");
+          g_string_append (json, g_ascii_dtostr (buf, sizeof (buf), point->x));
+          g_string_append (json, ", ");
+          g_string_append (json, g_ascii_dtostr (buf, sizeof (buf), point->y));
+        }
+
+      g_string_append (json, "]}");
+    }
+
+  g_string_append (json, "]");
+
+  return g_string_free (json, FALSE);
+}
+
+static void
+gimp_remove_tool_start (GimpRemoveTool *remove,
+                        GimpDisplay    *display)
+{
+  GimpRemoveOptions *options = GIMP_REMOVE_TOOL_GET_OPTIONS (remove);
+  GimpImage         *image   = gimp_display_get_image (display);
+  GList             *drawables;
+
+  drawables = gimp_image_get_selected_drawables (image);
+
+  if (! drawables)
+    {
+      gimp_tool_message_literal (GIMP_TOOL (remove), display,
+                                 _("Select the layer to remove from."));
+      return;
+    }
+
+  /*  the plug-in runs in a nested main loop, where the image can be
+   *  closed and the tool changed: run it once this event is done, with
+   *  the tool, image, display and drawable kept alive until it returns
+   */
+  remove->running      = TRUE;
+  remove->image        = g_object_ref (image);
+  remove->display      = g_object_ref (display);
+  remove->drawable     = g_object_ref (drawables->data);
+  remove->json         = gimp_remove_tool_strokes_json (remove);
+  remove->generative   = options->generative;
+
+  g_list_free (drawables);
+
+  g_idle_add_full (G_PRIORITY_DEFAULT_IDLE,
+                   gimp_remove_tool_run_idle,
+                   g_object_ref (remove), g_object_unref);
+}
+
+static gboolean
+gimp_remove_tool_run_idle (gpointer data)
+{
+  GimpRemoveTool *remove = GIMP_REMOVE_TOOL (data);
+  GimpImage      *image  = remove->image;
+  GimpValueArray *return_vals;
+  GError         *error  = NULL;
+
+  return_vals =
+    gimp_pdb_execute_procedure_by_name (image->gimp->pdb,
+                                        gimp_get_user_context (image->gimp),
+                                        GIMP_PROGRESS (remove->display),
+                                        &error,
+                                        REMOVE_PROC,
+                                        GIMP_TYPE_RUN_MODE, GIMP_RUN_NONINTERACTIVE,
+                                        GIMP_TYPE_IMAGE,    image,
+                                        GIMP_TYPE_DRAWABLE, remove->drawable,
+                                        G_TYPE_STRING,      remove->json,
+                                        G_TYPE_STRING,      remove->generative ?
+                                                            "klein" : "lama",
+                                        G_TYPE_NONE);
+
+  if (error)
+    {
+      /*  the plug-in's own message (no local AI, ComfyUI not answering...)  */
+      gimp_message_literal (image->gimp, G_OBJECT (remove->display),
+                            GIMP_MESSAGE_WARNING, error->message);
+      g_clear_error (&error);
+    }
+
+  gimp_value_array_unref (return_vals);
+  gimp_image_flush (image);
+
+  /*  the strokes are done (removed, or failed and reported)  */
+  gimp_draw_tool_pause (GIMP_DRAW_TOOL (remove));
+  g_ptr_array_set_size (remove->strokes, 0);
+  gimp_draw_tool_resume (GIMP_DRAW_TOOL (remove));
+
+  remove->running = FALSE;
+  g_clear_pointer (&remove->json, g_free);
+  g_clear_object (&remove->drawable);
+  g_clear_object (&remove->display);
+  g_clear_object (&remove->image);
+
+  return G_SOURCE_REMOVE;
+}
diff --git a/app/tools/gimpremovetool.h b/app/tools/gimpremovetool.h
new file mode 100644
index 0000000000..6529b54630
--- /dev/null
+++ b/app/tools/gimpremovetool.h
@@ -0,0 +1,65 @@
+/* GIMP - The GNU Image Manipulation Program
+ * Copyright (C) 1995 Spencer Kimball and Peter Mattis
+ *
+ * gimpremovetool.h
+ * Copyright (C) 2026 GIMPhoto contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "gimpdrawtool.h"
+
+
+#define GIMP_TYPE_REMOVE_TOOL            (gimp_remove_tool_get_type ())
+#define GIMP_REMOVE_TOOL(obj)            (G_TYPE_CHECK_INSTANCE_CAST ((obj), GIMP_TYPE_REMOVE_TOOL, GimpRemoveTool))
+#define GIMP_REMOVE_TOOL_CLASS(klass)    (G_TYPE_CHECK_CLASS_CAST ((klass), GIMP_TYPE_REMOVE_TOOL, GimpRemoveToolClass))
+#define GIMP_IS_REMOVE_TOOL(obj)         (G_TYPE_CHECK_INSTANCE_TYPE ((obj), GIMP_TYPE_REMOVE_TOOL))
+#define GIMP_IS_REMOVE_TOOL_CLASS(klass) (G_TYPE_CHECK_CLASS_TYPE ((klass), GIMP_TYPE_REMOVE_TOOL))
+#define GIMP_REMOVE_TOOL_GET_CLASS(obj)  (G_TYPE_INSTANCE_GET_CLASS ((obj), GIMP_TYPE_REMOVE_TOOL, GimpRemoveToolClass))
+
+#define GIMP_REMOVE_TOOL_GET_OPTIONS(t)  (GIMP_REMOVE_OPTIONS (gimp_tool_get_options (GIMP_TOOL (t))))
+
+
+typedef struct _GimpRemoveTool      GimpRemoveTool;
+typedef struct _GimpRemoveToolClass GimpRemoveToolClass;
+
+struct _GimpRemoveTool
+{
+  GimpDrawTool  parent_instance;
+
+  GimpVector2   last_pos;  /*  the brush outline, under the pointer      */
+  GPtrArray    *strokes;   /*  RemoveStroke: painted, not removed yet    */
+  gpointer      stroke;    /*  RemoveStroke being painted, or NULL       */
+  gboolean      running;   /*  the plug-in is removing (one at a time)   */
+
+  /*  the removal waiting for its idle run  */
+  GimpImage    *image;
+  GimpDisplay  *display;
+  GimpDrawable *drawable;
+  gchar        *json;
+  gboolean      generative;
+};
+
+struct _GimpRemoveToolClass
+{
+  GimpDrawToolClass  parent_class;
+};
+
+
+void    gimp_remove_tool_register (GimpToolRegisterCallback  callback,
+                                   gpointer                  data);
+
+GType   gimp_remove_tool_get_type (void) G_GNUC_CONST;
diff --git a/app/tools/meson.build b/app/tools/meson.build
index e48ca2c591..a0b80e5f0d 100644
--- a/app/tools/meson.build
+++ b/app/tools/meson.build
@@ -95,6 +95,8 @@ libapptools_sources = [
   'gimpshapeoptions.c',
   'gimpshapetool.c',
   'gimpobjectselecttool.c',
+  'gimpremoveoptions.c',
+  'gimpremovetool.c',
   'gimppenciltool.c',
   'gimpperspectiveclonetool.c',
   'gimpperspectivetool.c',
